### PR TITLE
wildcard ingress migration and webhook validation

### PR DESCRIPTION
**Pull request by joelddiaz**
- [x] don't allow the domain field on an ingress entry to lead with `*`. this gets copied over to the remote cluster, and the ingress controller will reject the leading `*` (the cluster's ingress controller assumes the domain is a wildcard).
- [x] ingress controller doesn't allow ingress entries where the ingress domain specified is different than the domain the cluster is under (eg cluster.example.com can't have an ingress like extra.notexample.com, it must be under cluster.example.com).
- [x] migrate old clusterdeployments with wildcard ingress entries to new format.

### CODE REVIEW

File: pkg/apis/hive/v1alpha1/validating-webhooks/clusterdeployment_validating_admission_hook.go [hunk 1]
```diff
@@ -455,6 +441,31 @@ func hasClearedOutPreviouslyDefinedIngressList(oldObject, newObject *hivev1.Clus
 	return false
 }
 
+func validateIngressDomainsShareClusterDomain(newObject *hivev1.ClusterDeploymentSpec) bool {
+	// ingress entries must share the same domain as the cluster
+	regexString := fmt.Sprintf("(?i).*%s.%s$", newObject.ClusterName, newObject.BaseDomain)
```
Comment (csrwng) [inline pkg/apis/hive/v1alpha1/validating-webhooks/clusterdeployment_validating_admission_hook.go]: what is the purpose of the ?i (for my own education)... also, shouldn't the regexp include a period before the cluster name?

Comment (joelddiaz) [reply pkg/apis/hive/v1alpha1/validating-webhooks/clusterdeployment_validating_admission_hook.go]: case insenstive.
yes, adding the period will improve the pattern matching. will add that.

File: pkg/controller/clusterdeployment/clusterdeployment_controller_test.go [hunk 2]
```diff
@@ -868,3 +893,68 @@ func testAvailableDNSZone() *hivev1.DNSZone {
...
+	return cd
+}
+
+func TestClusterDeploymentWildcardDomainMigration(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+
+	tests := []struct {
+		name              string
+		existing          *hivev1.ClusterDeployment
+		migrationExpected bool
```
[the remark this answers is not in the document]
Comment (joelddiaz) [reply pkg/controller/clusterdeployment/clusterdeployment_controller_test.go]: ok. added struct to allow listing any expected domains and what that string should look like for each defined ingress entry.

File: pkg/apis/hive/v1alpha1/validating-webhooks/clusterdeployment_validating_admission_hook.go (hunk 1 above)
Comment (joelddiaz) [reply pkg/apis/hive/v1alpha1/validating-webhooks/clusterdeployment_validating_admission_hook.go]: added the leading `.` as part of the shared domain validation

**Pull request merged**